backend/cfgize.ml

```diff
@@ -0,0 +1,808 @@
+    | Subf -> false
+    | Mulf -> false
+    | Divf -> false
+    | Compf _ -> false
+    | Floatofint -> false
+    | Intoffloat -> false
+    | Probe _ -> true
+    | Probe_is_enabled _ -> false (* CR xclerc for xclerc: double check *)
+    | Specific _ -> false (* CR xclerc for xclerc: double check *)
+    | Name_for_debugger _ -> false
+
+let can_raise_instr
+  : Cfg.basic Cfg.instruction -> bool
+  = fun instr ->
+    match instr.desc with
+    | Op op -> can_raise_operation op
+    | Call _ -> true
+    | Reloadretaddr -> false
+    | Pushtrap _ -> false
+    | Poptrap -> false
+    | Prologue -> false
+
+let can_raise_instrs
+  : Cfg.basic Cfg.instruction list -> bool
+  = fun instrs ->
+    List.exists can_raise_instr instrs
+
+let can_raise_terminator
+  : Cfg.terminator -> bool
+  = fun terminator ->
+    match terminator with
+    | Never -> false
+    | Always _ -> false
+    | Parity_test _ -> false
+    | Truth_test _ -> false
+    | Float_test _ -> false
+    | Int_test _ -> false
+    | Switch _ -> false
+    | Return -> false
+    | Raise _ -> true
+    | Tailcall (Self _) -> false
+    | Tailcall (Func _) -> true
+    | Call_no_return _ -> true
+
+let is_noop_move (instr : Cfg.basic Cfg.instruction) : bool =
+  match instr.Cfg.desc with
+  | Op (Move | Spill | Reload) ->
+    (* CR xclerc for xclerc: is testing the location enough? *)
+    Reg.same_loc instr.Cfg.arg.(0) instr.Cfg.res.(0)
+  | Op (Const_int _ | Const_float _ | Const_symbol _ | Stackoffset _
+       | Load _ | Store _ | Intop _ | Intop_imm _ | Negf | Absf | Addf
+       | Subf | Mulf | Divf | Compf _ | Floatofint | Intoffloat | Probe _
+       | Probe_is_enabled _ | Specific _ | Name_for_debugger _)
+  | Call _ | Reloadretaddr | Pushtrap _ | Poptrap | Prologue ->
+    false
+
+let rec get_end
+  : Mach.instruction -> Mach.instruction
+  = fun instr ->
+    match instr.Mach.desc with
+    | Iend -> instr
+    | Iop _
+    | Ireturn _
+    | Iifthenelse _
+    | Iswitch _
+    | Icatch _
+    | Iexit _
+    | Itrywith _
+    | Iraise _ ->
+      get_end instr.Mach.next
+
+type block_info = {
+  instrs : Cfg.basic Cfg.instruction list;
+  last : Mach.instruction;
+  terminator : Cfg.terminator Cfg.instruction option;
+  can_raise : bool;
+}
+
+(* [extract_block_info state first ~trap_depth] returns a [block_info]
+   containing all the instructions starting from [first] until some kind of
+   control flow is encountered or the end of the block is reached (i.e. [Iend]).
+   If the returned terminator is [None], it is guaranteed that the [last]
+   instruction is not an [Iop] (as it would either be part of [instrs] or be
+   a terminator). *)
+let extract_block_info
+  : State.t -> Mach.instruction -> trap_depth:int -> block_info
+  = fun state first ~trap_depth ->
+    let rec loop (instr : Mach.instruction) acc =
+      let return terminator can_raise =
+        let instrs =
+          List.rev (List.filter (fun instr -> not (is_noop_move instr)) acc) in
```

Comment:
Instead of  filtering out these NOP moves, it should be easy to avoid creating these instructions in the first place, but checking them in line 419-420.